FWCore/Utilities/interface/RandomNumberGenerator.h

```diff
@@ -171,7 +171,7 @@ namespace edm {
     /// the engine owned by the service and the cloned random engine should be
     /// used to generate numbers for initialization so that all initializations
     /// in the process get identical sequences of random numbers.
-    virtual std::unique_ptr<CLHEP::HepRandomEngine> cloneEngine(LuminosityBlockIndex const&) = 0;
+    virtual std::shared_ptr<CLHEP::HepRandomEngine> cloneEngine(LuminosityBlockIndex const&) = 0;
```

Comment:
I prefer to keep the `unique_ptr` interface here (to signify that the `cloneEngine()` itself does not hold shared ownership to the returned pointer).

You can construct a `shared_ptr` from `unique_ptr`. In general I'd recommend to minimize the use of `shared_ptr` to the minimum (i.e. in this case only where Pythia8 interface needs it, and the unavoidable knock-on effects of that), because shared ownership often makes the code more difficult to reason about e.g. on thread safety.

Reply:
Hi Matti, ok, thanks a lot, I will try to revert as much of the `shared_ptr` transition as possible!

Reply:
> Updated, please let me know on the new diff if you still spot some changes that should be rolled back

This part looks good now, thanks. I'd ask to squash the commits together eventually before this PR gets merged (to avoid back-and-forth changes in the git history).